synapse/dendrite/jobs.py

```diff
@@ -0,0 +1,279 @@
+
+    def fail(self, job):
+        '''
+        Fail the job provided.
+
+        The job provided is transitioned back to a state of 'queued', this will make the job available again for
+        retrieval from the queue.
+
+        Args:
+            job (dict):
+                The job to fail.
+        '''
+        tufo = self.core.getTufoByIden(job.get('iden'))
+        if tufo:
+            # for now, just log and requeue the job
+            logger.debug('Failing job %s', tufo)
+            self.core.setTufoProps(tufo, status='failed', failed_at=s_common.now())
+
+    def qsize(self, queue, status='queued'):
+        '''
+        Determine the size of the queue specified.
+
+        By default, the number of 'queued' jobs is returned. A status is optionally specified to determine the number
+        of jobs in other states, e.g. 'working' or 'completed'. The pseudo status of 'any' is also accepted for
+        counting all jobs regardless of status.
+
+        Args:
+            queue (str):
+                The name of the queue to determine the size of.
+
+            status (str):
+                The job status to predicate on while determining the size.
+
+        Returns:
+            int: Size of the queue.
+        '''
+        return len(self._jobsByQueue(queue, status))
+
+    def isEmpty(self, queue, status='queued'):
+        '''
+        Determine whether the queue specified is empty.
+
+        A convenience method for determining whether a queue's size is 0.
+
+        Args:
+            queue (str):
+                The name of the queue to determine the size of.
+
+            status (str):
+                The job status to predicate on while determining if a queue is empty.
+
+        Returns:
+            bool: True if the queue is empty, False otherwise.
+        '''
+        return self.qsize(queue, status) == 0
+
+    def stats(self, status='any'):
+        '''
+        Returns a histogram showing job count by queue name.
+
+        The default status argument is 'any', which is a pseudo status for collecting all jobs regardless of status.
+
+        Args:
+            status (str):
+                The status to filter the jobs by, e.g. 'queued', 'working', 'completed'.
+
+        Returns:
+            collections.defaultdict: A dictionary containing a histogram of count by queue name.
+        '''
+        if status == 'any':
+            return self.core.getStatByProp('histo', 'dendrite:job:queue')
+
+        histo = collections.defaultdict(int)
+        for tufo in self.core.getTufosByProp('dendrite:job:status', status):
+            histo[tufo[1].get('dendrite:job:queue')] += 1
+        return histo
+
+    def _initJobProps(self, queue, job):
+        return {
+            'queue':    queue,
+            'data':     json.dumps(job),
+            'runkey':   self._runkey(),
+            'status':   'queued',
+            'queued_at': s_common.now()
+        }
+
+    def _runkey(self):
```

Comment:
this probably needs a lock to avoid race conditions from multiple simultaneous callers. If a thread for caller A were scheduled out (even briefly) after line 249, and caller B were to come through where currMs had changed, jobNum would be smashed to 1 and then used by both.